File: CaptureGo/board.py
from move import Move
import random
import numpy as np
from group import Group
import copy


class Board:
    EMPTY = 0
    BLACK = 1
    WHITE = -1
    BORDER = 3
    def __init__(self,size,adjacency_table = None):
        self.size=size
        self.occupied = set()
        self.grid = {}
        self.adjacency_table = adjacency_table
        if self.adjacency_table is None:
            self.adjacency_table = create_adjacency_table(self.size)



    def move(self,move):
        point = (move.x, move.y)
        self.occupied.add(point)
        
        new_group = self.create_stone(move)
        adjacent_groups = self.adjacent_groups(point)
        for group in adjacent_groups: 
            for stone in group.stones:
                self.grid[stone] = self.grid[stone].remove_liberty(point)
    

        adjacent_groups = self.adjacent_groups(point)
        new_group = self.create_stone(move)
        for group in adjacent_groups:
            if group.color == move.turn:
                new_group = new_group.merge_group(group)

        for stone in new_group.stones:
            self.grid[stone] = new_group
        

    def create_stone(self,move):
        adjacent_lib = self.adjacent_liberties(move)
        group = Group(move.turn,[(move.x, move.y)],adjacent_lib)
        return group

    def adjacent_liberties(self,move):
        return self.adjacency_table[(move.x,move.y)]-self.occupied

    def adjacent_groups(self,point):
        adjacent_groups = set()
        adjacent_stones = self.adjacent_stones(point)
        for stone in adjacent_stones:
                adjacent_groups.add(self.grid[stone])
        return adjacent_groups
    
    def adjacent_stones(self,point):
        return self.adjacency_table[point] & self.occupied

    def adjacent_groups_index(self,point):
        adjacent_group_index = []
        for i in range(len(self.groups)):
            if point in self.groups[i].liberties:
                if not i in adjacent_group_index:   
                    adjacent_group_index.append(i)
        return adjacent_group_index

    def is_legal_move(self,move):
        if (move.x, move.y) in self.occupied:
            return False
        elif self.is_capture(move):
            return True
        elif self.is_self_capture(move):
            return False
        else:
            return True
    
    #assuming move isn't a capture it checks if it is a self capture
    def is_self_capture(self,move):
        group = self.create_stone(move)
        
        if group.number_of_liberties() > 0:
            return False
        else:
            adjacent_groups = self.adjacent_groups((move.x, move.y))
            for group in adjacent_groups:
                if group.color == move.turn:
                    if group.number_of_liberties() >= 2:
                        return False
        return True

    
    def is_capture(self,move):
        ENEMY = -move.turn
        adjacent_groups = self.adjacent_groups((move.x, move.y))
        for group in adjacent_groups:
            if group.color == ENEMY:
                if group.number_of_liberties() == 1:
                    return True
        return False

    def print_board(self):
        np_board= np.zeros(((self.size),(self.size)),dtype=np.float32)
        for i in range(1,self.size+1):
            for j in range(1,self.size+1):
                point= (i,j)
                if point in self.occupied:
                    np_board[i-1][j-1] = self.grid[(point)].color
        

        for i in range(0,self.size):
            for j in range(0,self.size):
                color = np_board[j][i]
                if color == self.EMPTY:
                    print(".",end=" ")
                elif color == self.BLACK:
                    print("B",end=" ")
                else:
                    print("W",end=" ")    
            print("")
        


    #returns a list of all legal moves for a given color
    def legal_moves(self,color):
        moves = []
        candidates = self.adjacency_table.keys() - self.occupied
        for candidate in candidates:
            move = Move(color,candidate[0],candidate[1])
            if self.is_legal_move(move):
                moves.append(move)
        return moves

        
    def __deepcopy__(self,memo={}):
        board = Board(self.size,adjacency_table=self.adjacency_table)
        board.occupied = copy.copy(self.occupied)
        board.grid = copy.copy(self.grid) #shallow copy
        return board
        

def create_adjacency_table(size):
    adjacency_table = {}
    
    #inside points
    for i in range(2,size):
        for j in range(2,size):
            point = (i, j)
            adjacency_table[point] = {(i-1,j),(i+1,j),(i,j-1),(i,j+1)}
    #edge points
    for i in range (2,size):
        adjacency_table[(1,i)] = {(2,i),(1,i-1),(1,i+1)}
        adjacency_table[(size,i)] = {(size-1,i),(size,i-1),(size,i+1)}
        adjacency_table[(i,1)] = {(i,2),(i-1,1),(i+1,1)}
        adjacency_table[(i,size)] = {(i,size-1),(i-1,size),(i+1,size)}
    #corner points
    adjacency_table[(1,1)] = {(1,2),(2,1)}
    adjacency_table[(1,size)] = {(2,size),(1,size-1)}
    adjacency_table[(size,size)] = {(size,size-1),(size-1,size)}
    adjacency_table[(size,1)] = {(size,2),(size-1,1)}
    return adjacency_table

File: SelfplayEngine/gameplay/board.cc
#include "board.h"
#include <stdio.h>




bool Board::is_legal(Move move){
    if (groups.count({move.x, move.y})){
        return false;
    }
    else if(is_capture(move)){
        return true;
    }
    else if(is_self_capture(move)){
        return false;
    }
    else{
        return true;
    }
}

void Board::play_move(Move move){
    Point point({move.x,move.y});
    std::unordered_set<Point> s = {point};
    std::unordered_set<Point> l = adjacency_table->at(point);

    std::shared_ptr<Group> group = std::make_shared<Group>(move.turn,s,l);
    
    for (auto neighbor : l){
        if (groups.count(neighbor)){
            group->remove_liberty(neighbor);
            std::shared_ptr<Group> neighbor_group = groups.at(neighbor);
            neighbor_group->remove_liberty(point);
            groups.at(neighbor)->remove_liberty(point);
            //merge adjacent groups of same color
            if(neighbor_group->color == move.turn){
                group->merge_group(*neighbor_group);
                groups.at(neighbor) = group;
            }
        }
    }
    //update the group of the point
    groups.insert({point,group});
    //update
    for (auto stone : group->stones){
        groups.at(stone) = group;
    }
}

bool Board::is_capture(Move move){
    std::unordered_set<Point> neighbors = adjacency_table->at({move.x, move.y});
    for (auto neighbor: neighbors){
        if (groups.count(neighbor)){
            std::shared_ptr<Group> group = groups.at(neighbor);
            if (group->color != move.turn){
                if (group->number_of_liberties() == 1){
                    return true;
                }
            }
        }
    }
    return false;
}

bool Board::is_self_capture(Move move){
    std::unordered_set<Point> neighbors = adjacency_table->at({move.x, move.y});
    std::unordered_set<Point> s = {{move.x, move.y}};
    std::unordered_set<Point> l = adjacency_table->at({move.x, move.y});
    std::shared_ptr<Group> group = std::make_shared<Group>(move.turn,s,l);

    for (auto neighbor: neighbors){
        if (groups.count(neighbor)){
            group->remove_liberty(neighbor);
            std::shared_ptr<Group> neighbour_group = groups.at(neighbor);
            if(neighbour_group->color == move.turn){
                group->merge_group(*neighbour_group);
            }
        }
    }
    if(group->number_of_liberties() == 0){
        return true;
    }
    else{
        return false;
    }
}

std::vector<Point> Board::legal_moves(int color){
    std::vector<Point> moves;
    for (int i = 0; i < size; i++){
        for (int j = 0; j < size; j++){
            Move move(color,i,j);
            if (is_legal(move)){
                moves.push_back({i,j});
            }
        }
    }
    return moves;
}

void Board::print_board(){
    for (int i = 0; i < size; i++){
        for (int j = 0; j < size; j++){
            if(groups.count({i,j})){
                if(groups.at({i,j})->color == 1){
                    printf("B");
                }
                else{
                    printf("W");
                }
            }
            else{
                printf(".");
            }
        }
        printf("\n");
    }  
}



void Board::create_adjacency_table(){
    std::cout<<"Creating adjacency table\n";
    // Inside points
    for (int i = 1; i < size-1; i++){
        for (int j = 1; j < size-1; j++){
            adjacency_table->insert({{i,j},{{i-1, j}, {i+1, j}, {i, j-1}, {i, j+1}}});
        }
    }
    
    // Edge points
    for (int i = 1; i < size-1; i++){
        adjacency_table->insert({{0, i},{{1, i}, {0, i-1}, {0, i+1}}});
        adjacency_table->insert({{size-1, i},{{size-2, i}, {size-1, i-1}, {size-1, i+1}}});
        adjacency_table->insert({{i, 0},{{i, 1}, {i-1, 0}, {i+1, 0}}});
        adjacency_table->insert({{i, size-1},{{i, size-2}, {i-1, size-1}, {i+1, size-1}}});
    }
    
    // Corner points
    adjacency_table->insert({{0, 0},{{0, 1}, {1, 0}}});
    adjacency_table->insert({{0, size-1},{{1, size-1}, {0, size-2}}});
    adjacency_table->insert({{size-1, size-1},{{size-2, size-1}, {size-1, size-2}}});
    adjacency_table->insert({{size-1, 0},{{size-1, 1}, {size-2, 0}}});
}